README.md
This is my personal AI for Screeps (https://screeps.com/), an open source MMO RTS sandbox game wherein the core mechanic is programming your custom AI.


changelog.txt
0.11
- Second room controller

0.10
- Second room expansion

0.9
- Remote mining reenabled

0.8
- Minor fixes

0.7
- Creep functions refactored

0.6
- Harvester fix
- Body parts refactored

0.5
- Functions now belong to system objects

0.4
- Moving with Claudio

0.3
- Room expansion

0.2
- External room mining
- Readme added

0.1
- Basic behaviour for creeps, spawn, towers and remote mining


body.js
module.exports = {

    harvester: [WORK,WORK,CARRY,MOVE],

    claimer: [CLAIM,WORK,WORK,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE],

    miner: [
        [MOVE], // 0
        [MOVE], // 1
        [WORK,WORK,WORK,CARRY,MOVE], // 2
        [WORK,WORK,WORK,WORK,WORK,CARRY,MOVE], // 3
        [WORK,WORK,WORK,WORK,WORK,WORK,WORK,CARRY,MOVE], // 4
        [WORK,WORK,WORK,WORK,WORK,WORK,WORK,CARRY,MOVE,MOVE] // 5
    ],

    charger: [
        [MOVE], // 0
        [MOVE], // 1
        [CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE], // 2
        [CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE], // 3
        [CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE], // 4
        [CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE] // 5
    ],

    hauler: [
        [MOVE], // 0
        [MOVE], // 1
        [CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE], // 2
        [CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE], // 3
        [CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE], // 4
        [CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE] // 5
    ],

    upgrader: [
        [WORK,WORK,CARRY,MOVE], // 0
        [WORK,WORK,CARRY,MOVE], // 1
        [WORK,WORK,WORK,CARRY,MOVE], // 2
        [WORK,WORK,WORK,WORK,WORK,CARRY,MOVE], // 3
        [WORK,WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE], // 4
        [WORK,WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE,MOVE] // 5
    ],

    engineer: [
        [WORK,WORK,CARRY,MOVE], // 0
        [WORK,WORK,CARRY,MOVE], // 1
        [WORK,WORK,WORK,CARRY,MOVE], // 2
        [WORK,WORK,WORK,WORK,WORK,CARRY,MOVE], // 3
        [WORK,WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE], // 4
        [WORK,WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE,MOVE] // 5
    ],

    architect: [
        [MOVE], // 0
        [MOVE], // 1
        [MOVE], // 2
        [WORK,WORK,WORK,WORK,WORK,CARRY,MOVE], // 3
        [WORK,WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE], // 4
        [WORK,WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE] // 5
    ],

    explorer: [
        [MOVE], // 0
        [MOVE], // 1
        [MOVE], // 2
        [WORK,WORK,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE], // 3
        [WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE], // 4
        [WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE] // 5
    ]
};


main.js
//require('prototype.spawn')();
require('prototype.spawn');
require('prototype.creep');
require('prototype.tower');

// room.energyAvailable
//room.energyCapacityAvailable

module.exports.loop = function () {

    // Update memory once in a while
    if(Game.time % 20 === 0) {
        for (let name in Memory.creeps) {
            if (Game.creeps[name] == undefined) {
                delete Memory.creeps[name];
            }
        }
    }

    // Update spawns
    for(let name in Game.spawns) {
         Game.spawns[name].update();
    }

    // Update creeps
    for(let name in Game.creeps) {
        Game.creeps[name].update();
        //creepExt.update(creep);
        //Memory.room[creep.room.name].containerA = ['5d8406e2ec18233c4ea3311c', '5d8407fee9930148a560f2d7'];
        //Memory.room[creep.room.name].containerA = ['5d85686dbcb9d00d2569e637'];
    }

    // Update towers
    var towers = _.filter(Game.structures, s => s.structureType == STRUCTURE_TOWER);
    for(let tower of towers) {
        tower.update();
    }
};


prototype.spawn.js
var body = require('body');
const level = [ 300, 400, 600, 800, 1000, 1500 ];
var nextLevel = level[1];

// Updates spawn every tick
StructureSpawn.prototype.update =
    function() {
        // Updates spawn level
        this.updateLevel();

        // Spawns creeps when needed
        this.spawnCreeps();
    };

// Updates spawner level depending on the room's total energy capacity available
StructureSpawn.prototype.updateLevel =
    function() {
        this.memory.level = 0;
        if(this.room.energyCapacityAvailable > level[0] && this.room.energyCapacityAvailable < level[1]) {
            this.memory.level = 1;
        } else if(this.room.energyCapacityAvailable > level[1] && this.room.energyCapacityAvailable < level[2]) {
            this.memory.level = 2;
        } else if(this.room.energyCapacityAvailable >= level[2] && this.room.energyCapacityAvailable < level[3]) {
            this.memory.level = 3;
        } else if(this.room.energyCapacityAvailable >= level[3] && this.room.energyCapacityAvailable < level[4]) {
            this.memory.level = 4;
        } else if(this.room.energyCapacityAvailable >= level[4] && this.room.energyCapacityAvailable < level[5]) {
            this.memory.level = 5;
        } else if(this.room.energyCapacityAvailable >= level[5]) {
            this.memory.level = 5;
        }
    };

// Spawns creeps when needed
StructureSpawn.prototype.spawnCreeps =
    function() {

        // Count current creeps
        var currentHarvesters = _.sum(Game.creeps, (c) => c.memory.role == 'harvester');
        var currentMiners = _.sum(Game.creeps, (c) => c.memory.role == 'miner');
        var currentChargers = _.sum(Game.creeps, (c) => c.memory.role == 'charger');
        var currentUpgraders = _.sum(Game.creeps, (c) => c.memory.role == 'upgrader');
        var currentEngineers = _.sum(Game.creeps, (c) => c.memory.role == 'engineer');
        var currentArchitects = _.sum(Game.creeps, (c) => c.memory.role == 'architect');
        var currentExplorers = _.sum(Game.creeps, (c) => c.memory.role == 'explorer');
        var currentClaimers = _.sum(Game.creeps, (c) => c.memory.role == 'claimer');
        var currentHaulers = _.sum(Game.creeps, (c) => c.memory.role == 'hauler');

        // Calculate minimum creeps
        var minHarvesters = 0;
        var minMiners = 0;
        var minChargers = 0;
        var minUpgraders = 0;
        var minEngineers = 0;
        var minArchitects = 0;
        var minExplorers = 0;
        var minClaimers = 0;
        var minHaulers = 0;
        switch(this.memory.level) {
            case 0: // 300
            case 1:
                minHarvesters = 4;
                minUpgraders = 8;
                minEngineers = 8;
                break;

            case 2: // 400
                minMiners = 3;
                minChargers = 2;
                minUpgraders = 4;
                minEngineers = 6;
                break;

            case 3: // 600
                minMiners = 4;
                minChargers = 2;
                minUpgraders = 4;
                minEngineers = 5;
                minExplorers = 0;
                minArchitects = 1;
                break;

            case 4: // 800
                minMiners = 4;
                minChargers = 1;
                minUpgraders = 4;
                minEngineers = 1;
                minExplorers = 0;
                minArchitects = 2;
                break;

            case 5: // 1000
                minMiners = 2;
                minChargers = 1;
                minUpgraders = 3;
                minEngineers = 6;
                minExplorers = 5;
                minArchitects = 1;
                minHaulers = 0;
                minClaimers = 0;
                break;
        }

        // Create creeps
        if(this.memory.level < 2) {
            if(currentHarvesters < minHarvesters) {
                name = this.createCreep(body.harvester, undefined, { role: 'harvester', working: false, target: this.room.name, home: this.room.name});
            } else if(currentUpgraders < minUpgraders) {
                name = this.createCreep(body.upgrader[this.memory.level], undefined, { role: 'upgrader', working: false, target:  this.room.name, home: this.room.name});
            } else if(currentEngineers < minEngineers) {
                name = this.createCreep(body.engineer[this.memory.level], undefined, { role: 'engineer', working: false, target: this.room.name, home: this.room.name});
            } else {
                name = this.createCreep(body.upgrader[this.memory.level], undefined, { role: 'upgrader', working: false, target:  this.room.name, home: this.room.name});
            }
        } else {
            if(currentMiners == 0) {
                if(currentHarvesters == 0) {
                    name = this.createCreep(body.harvester, undefined, { role: 'harvester', working: false, target: this.room.name, home: this.room.name});
                } else {
                    name = this.createCreep(body.miner[this.memory.level], undefined, { role: 'miner', working: false, target: this.room.name, home: this.room.name});
                }
            } else if(currentMiners == 1 && currentChargers == 0) {
                name = this.createCreep(body.charger[this.memory.level], undefined, { role: 'charger', working: false, target: this.room.name, home: this.room.name});
            } else if(currentMiners < minMiners) {
                name = this.createCreep(body.miner[this.memory.level], undefined, { role: 'miner', working: false, target: this.room.name, home: this.room.name});
            } else if(currentChargers < minChargers) {
                name = this.createCreep(body.charger[this.memory.level], undefined, { role: 'charger', working: false, target: this.room.name, home: this.room.name});
            } else if(currentUpgraders < minUpgraders) {
                name = this.createCreep(body.upgrader[this.memory.level], undefined, { role: 'upgrader', working: false, target: "W21N47", home: this.room.name});
            } else if(currentExplorers < minExplorers) {
                name = this.createCreep(body.explorer[this.memory.level], undefined, { role: 'explorer', working: false, target: "W21N47", home: this.room.name});
            } else if(currentHaulers < minHaulers) {
                name = this.createCreep(body.hauler[this.memory.level], undefined, { role: 'hauler', working: false, target: this.room.name, home: this.room.name});
            } else if(currentEngineers < minEngineers) {
                name = this.createCreep(body.engineer[this.memory.level], undefined, { role: 'engineer', working: false, target: this.room.name, home: this.room.name});
            } else if(currentArchitects < minArchitects) {
                name = this.createCreep(body.architect[this.memory.level], undefined, { role: 'architect', working: false, target: this.room.name, home: this.room.name});
            } else if(currentClaimers < minClaimers) {
                name = this.createCreep(body.claimer, undefined, { role: 'claimer', working: false, target: "W21N47", home: this.room.name});
            }
        }

        // Prints console log every tick
        console.log(this.name + " lvl " + this.memory.level + " | Energy: " + this.room.energyAvailable + "/" + this.room.energyCapacityAvailable + " | CPU: " + Game.cpu.getUsed().toFixed(2))
        console.log(currentHarvesters + " harvesters, " + currentMiners + " miners, " + currentChargers + " chargers, "
        + currentUpgraders + " upgraders, " + currentEngineers + " engineers, " + currentArchitects + " architects, " + currentExplorers + " explorers, " + currentClaimers + " claimers, " + currentHaulers + " haulers.");
        console.log("-----------------------------------------------------------------------------------------------------");

    };


spawnExt.js
const level = [ 300, 400, 600, 800, 1000, 1500 ];
var nextLevel = level[1];

module.exports = {

    // Updates spawn every tick
    update: function(spawn) {

        // Updates spawn level
        this.updateLevel(spawn);

        // Spawns creeps when needed
        this.spawnCreeps(spawn);
    },

    // Updates spawner level depending on the room's total energy capacity available
    updateLevel: function(spawn) {

        // Updates spawner level
        spawn.memory.level = 0;
        if(spawn.room.energyCapacityAvailable > level[0] && spawn.room.energyCapacityAvailable < level[1]) {
            spawn.memory.level = 1;
        } else if(spawn.room.energyCapacityAvailable > level[1] && spawn.room.energyCapacityAvailable < level[2]) {
            spawn.memory.level = 2;
        } else if(spawn.room.energyCapacityAvailable >= level[2] && spawn.room.energyCapacityAvailable < level[3]) {
            spawn.memory.level = 3;
        } else if(spawn.room.energyCapacityAvailable >= level[3] && spawn.room.energyCapacityAvailable < level[4]) {
            spawn.memory.level = 4;
        } else if(spawn.room.energyCapacityAvailable >= level[4] && spawn.room.energyCapacityAvailable < level[5]) {
            spawn.memory.level = 5;
        } else if(spawn.room.energyCapacityAvailable >= level[5]) {
            spawn.memory.level = 5;
        }
    },

    // Spawns creeps when needed
    spawnCreeps: function(spawn) {

        var minHarvesters = 0;
        var minMiners = 0;
        var minChargers = 0;
        var minUpgraders = 0;
        var minEngineers = 0;
        var minArchitects = 0;
        var minExplorers = 0;
        var minClaimers = 0;
        var minHaulers = 0;

        // Count current active roles
        var currentHarvesters = _.sum(Game.creeps, (c) => c.memory.role == 'harvester');
        var currentMiners = _.sum(Game.creeps, (c) => c.memory.role == 'miner');
        var currentChargers = _.sum(Game.creeps, (c) => c.memory.role == 'charger');
        var currentUpgraders = _.sum(Game.creeps, (c) => c.memory.role == 'upgrader');
        var currentEngineers = _.sum(Game.creeps, (c) => c.memory.role == 'engineer');
        var currentArchitects = _.sum(Game.creeps, (c) => c.memory.role == 'architect');
        var currentExplorers = _.sum(Game.creeps, (c) => c.memory.role == 'explorer');
        var currentClaimers = _.sum(Game.creeps, (c) => c.memory.role == 'claimer');
        var currentHaulers = _.sum(Game.creeps, (c) => c.memory.role == 'hauler');

        var name = undefined;
        switch(spawn.memory.level) {

            case 0: // 300
            case 1:

                minHarvesters = 4;
                minUpgraders = 8;
                minEngineers = 8;

                // harvester - upgrader - repairer - engineer - upgrader
                if(currentHarvesters < minHarvesters) {
                    name = spawn.createCreep([WORK,WORK,CARRY,MOVE], undefined, { role: 'harvester', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentUpgraders < minUpgraders) {
                    name = spawn.createCreep([WORK,WORK,CARRY,MOVE], undefined, { role: 'upgrader', working: false, level: level, target:  spawn.room.name, home: spawn.room.name});
                } else if(currentEngineers < minEngineers) {
                    name = spawn.createCreep([WORK,WORK,CARRY,MOVE], undefined, { role: 'engineer', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else {
                    name = spawn.createCreep([WORK,WORK,CARRY,MOVE], undefined, { role: 'upgrader', working: false, level: level, target:  spawn.room.name, home: spawn.room.name});
                }
                break;

            case 2: // 400

                minMiners = 3;
                minChargers = 2;
                minUpgraders = 4;
                minEngineers = 6;

                // harvester - miner - charger - upgrader - engineer - upgrader
                if(currentMiners == 0) {
                    if(currentHarvesters == 0) {
                        name = spawn.createCreep([WORK,WORK,CARRY,MOVE], undefined, { role: 'harvester', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                    } else {
                        name = spawn.createCreep([WORK,WORK,WORK,CARRY,MOVE], undefined, { role: 'miner', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                    }
                } else if(currentMiners == 1 && currentChargers == 0) {
                    name = spawn.createCreep([CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'charger', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentMiners < minMiners) {
                    name = spawn.createCreep([WORK,WORK,WORK,CARRY,MOVE], undefined, { role: 'miner', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentChargers < minChargers) {
                    name = spawn.createCreep([CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'charger', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentUpgraders < minUpgraders) {
                    name = spawn.createCreep([WORK,WORK,WORK,CARRY,MOVE], undefined, { role: 'upgrader', working: false, level: level, target:  spawn.room.name, home: spawn.room.name});
                } else if(currentEngineers < minEngineers) {
                    name = spawn.createCreep([WORK,WORK,WORK,CARRY,MOVE], undefined, { role: 'engineer', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else {
                    name = spawn.createCreep([WORK,WORK,WORK,CARRY,MOVE], undefined, { role: 'upgrader', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                }
                break;

            case 3: // 600
                minMiners = 4;
                minChargers = 2;
                minUpgraders = 4;
                minEngineers = 5;
                minExplorers = 0;
                minArchitects = 1;

                // harvester - miner - charger - upgrader - engineer - upgrader
                if(currentMiners == 0) {
                    if(currentHarvesters == 0) {

                        name = spawn.createCreep([WORK,WORK,CARRY,MOVE], undefined, { role: 'harvester', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                    } else {
                        name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,CARRY,MOVE], undefined, { role: 'miner', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                    }
                } else if(currentMiners == 1 && currentChargers == 0) {
                    name = spawn.createCreep([CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'charger', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentMiners < minMiners) {
                    name = spawn.createCreep([WORK,WORK,WORK,WORK,CARRY,MOVE,MOVE], undefined, { role: 'miner', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentChargers < minChargers) {
                    name = spawn.createCreep([CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'charger', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentUpgraders < minUpgraders) {
                    name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,CARRY,MOVE], undefined, { role: 'upgrader', working: false, level: level, target:  spawn.room.name, home: spawn.room.name});
                } else if(currentEngineers < minEngineers) {
                    name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,CARRY,MOVE], undefined, { role: 'engineer', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentArchitects < minArchitects) {
                    name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,CARRY,MOVE], undefined, { role: 'architect', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentExplorers < minExplorers) {
                    name = spawn.createCreep([WORK,WORK,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE], undefined, { role: 'explorer', working: false, level: level, target: spawn.room.name, home: spawn.room.name, target: spawn.room.name, home: spawn.room.name});
                } else {
                    name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,CARRY,MOVE], undefined, { role: 'upgrader', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                }
                break;

            case 4: // 800
                minMiners = 4;
                minChargers = 1;
                minUpgraders = 4;
                minEngineers = 1;
                minExplorers = 0;
                minArchitects = 2;

                // harvester - miner - charger - upgrader - engineer - upgrader
                if(currentMiners == 0) {
                    if(currentHarvesters == 0) {

                        name = spawn.createCreep([WORK,WORK,CARRY,MOVE], undefined, { role: 'harvester', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                    } else {
                        name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,WORK,WORK,CARRY,MOVE], undefined, { role: 'miner', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                    }
                } else if(currentMiners == 1 && currentChargers == 0) {
                    name = spawn.createCreep([CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'charger', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentMiners < minMiners) {
                    name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,WORK,WORK,CARRY,MOVE], undefined, { role: 'miner', working: false, level: level, target: 'E16N31', home: spawn.room.name});
                } else if(currentChargers < minChargers) {
                    name = spawn.createCreep([CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'charger', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentUpgraders < minUpgraders) {
                    name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE], undefined, { role: 'upgrader', working: false, level: level, target:  spawn.room.name, home: spawn.room.name});
                } else if(currentEngineers < minEngineers) {
                    name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE], undefined, { role: 'engineer', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentArchitects < minArchitects) {
                    name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE], undefined, { role: 'architect', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentExplorers < minExplorers) {
                    name = spawn.createCreep([WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE], undefined, { role: 'explorer', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                }
                break;

            case 5: // 1000
                minMiners = 3;
                minChargers = 1;
                minUpgraders = 4;
                minEngineers = 2;
                minExplorers = 0;
                minArchitects = 1;
                minHaulers = 0;
                minClaimers = 0;

                // harvester - miner - charger - upgrader - engineer - upgrader
                if(currentMiners == 0) {
                    if(currentHarvesters == 0) {
                        name = spawn.createCreep([WORK,WORK,CARRY,MOVE], undefined, { role: 'harvester', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                    } else {
                        name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,WORK,WORK,CARRY,MOVE,MOVE], undefined, { role: 'miner', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                    }
                } else if(currentMiners == 1 && currentChargers == 0) {
                    name = spawn.createCreep([CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'charger', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentMiners < minMiners) {
                    name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,WORK,WORK,CARRY,MOVE,MOVE], undefined, { role: 'miner', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentChargers < minChargers) {
                    name = spawn.createCreep([CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'charger', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentHaulers < minHaulers) {
                    name = spawn.createCreep([CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'hauler', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentUpgraders < minUpgraders) {
                    name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE], undefined, { role: 'upgrader', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentEngineers < minEngineers) {
                    name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE], undefined, { role: 'engineer', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentArchitects < minArchitects) {
                    name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE], undefined, { role: 'architect', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentExplorers < minExplorers) {
                    if(spawn.memory.flag == 0) {
                        name = spawn.createCreep([WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'explorer', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                        spawn.memory.flag = 1;
                    } else {
                        name = spawn.createCreep([WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'explorer', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                        spawn.memory.flag = 0;
                    }
                } else if(currentClaimers < minClaimers) {
                    name = spawn.createCreep([CLAIM,WORK,WORK,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'claimer', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                }// else if(currentUpgraders < (minUpgraders + 1)) { // External upgraders (dirty)
                //    name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'upgrader', working: false, level: level, target: 'E16N31', home: 'E16N31'});
                //}
                break;

            case 6: // 1500
                minMiners = 3;
                minChargers = 1;
                minUpgraders = 4;
                minEngineers = 1;
                minExplorers = 0;
                minArchitects = 1;
                minHaulers = 0;
                minClaimers = 0;

                // harvester - miner - charger - upgrader - engineer - upgrader
                if(currentMiners == 0) {
                    if(currentHarvesters == 0) {
                        name = spawn.createCreep([WORK,WORK,CARRY,MOVE], undefined, { role: 'harvester', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                    } else {
                        name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,WORK,WORK,CARRY,MOVE,MOVE], undefined, { role: 'miner', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                    }
                } else if(currentMiners == 1 && currentChargers == 0) {
                    name = spawn.createCreep([CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'charger', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentMiners < minMiners) {
                    name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,WORK,WORK,CARRY,MOVE,MOVE], undefined, { role: 'miner', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentChargers < minChargers) {
                    name = spawn.createCreep([CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'charger', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentHaulers < minHaulers) {
                    name = spawn.createCreep([CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'hauler', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentUpgraders < minUpgraders) {
                    name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,WORK,WORK,WORK,WORK,WORK,CARRY,CARRY,MOVE,MOVE,MOVE], undefined, { role: 'upgrader', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentEngineers < minEngineers) {
                    name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE], undefined, { role: 'engineer', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentArchitects < minArchitects) {
                    name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE], undefined, { role: 'architect', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                } else if(currentExplorers < minExplorers) {
                    if(spawn.memory.flag == 0) {
                        name = spawn.createCreep([WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'explorer', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                        spawn.memory.flag = 1;
                    } else {
                        name = spawn.createCreep([WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'explorer', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                        spawn.memory.flag = 0;
                    }
                } else if(currentClaimers < minClaimers) {
                    name = spawn.createCreep([CLAIM,WORK,WORK,CARRY,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'claimer', working: false, level: level, target: spawn.room.name, home: spawn.room.name});
                }// else if(currentUpgraders < (minUpgraders + 1)) { // External upgraders (dirty)
                //    name = spawn.createCreep([WORK,WORK,WORK,WORK,WORK,WORK,CARRY,CARRY,CARRY,MOVE,MOVE,MOVE,MOVE,MOVE], undefined, { role: 'upgrader', working: false, level: level, target: 'E16N31', home: 'E16N31'});
                //}
                break;
        }

        // Prints console log every tick
        console.log(spawn.name + " level " + spawn.memory.level + " | Energy: " + spawn.room.energyAvailable + "/" + spawn.room.energyCapacityAvailable)
        console.log(currentHarvesters + " harvesters, " + currentMiners + " miners, " + currentChargers + " chargers, "
        + currentUpgraders + " upgraders, " + currentEngineers + " engineers, " + currentArchitects + " architects, " + currentExplorers + " explorers, " + currentClaimers + " claimers, " + currentHaulers + " haulers.");
        console.log("-----------------------------------------------------------------------------------------------------");

    },

    // Adds a room to the upgrade target room list in the spawn memory
    upgradeRoom: function(spawn, target) {
        if(!spawn.memory.upgrade) {
            spawn.memory.upgrade = [];
        }
        spawn.memory.upgrade.push(target);
    }
};
